### DISCUSSION TITLE
Binary sensor for Doorbell button with "Pressed" and "Released" states

### DISCUSSION THREAD

**Discussion by iCSpotRun** [uilibs/uiprotect#37] [Ideas]
### Is your feature request related to a problem? Please describe.

Currently HA users only have access to the "last_doorbell_ring" sensor, and the only available states are "Unavailable" and "Unknown".

My doorbell automation sends chimes through my house speakers. It's set to trigger on any state change. This works when the doorbell button is pressed, but the state sometimes changes randomly without apparent reason, causing doorbell chimes at odd hours like 3:30am. Not at all wife approved.

I'm wondering if another sensor can be added that supports "Pressed" and "Released" states.

### Describe alternatives you've considered

I'm not aware of alternatives, but open to using a workaround if someone has one in mind.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Are you willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time, and I know how to start.
- [X] Yes, I have the time, but I don't know how to start. I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by bdraco** [comment]
> Currently HA users only have access to the "last_doorbell_ring" sensor, and the only available states are "Unavailable" and "Unknown".

That kinda seems like a bug? Is there an open issue in HA issues?

**Comment by bdraco** [comment]
> I'm wondering if another sensor can be added that supports "Pressed" and "Released" states.
> 
> ###

The doorbell should now be implemented as an event entity.  The event entity was not available when it was originally added to Home Assistant

The integration is looking for a maintainer, and I’m just helping out to keep it warm until someone can be found, so we have to wait for them to take a look at that unless it's something you would like to add to HA

**Comment by iCSpotRun** [comment]
> > Currently HA users only have access to the "last_doorbell_ring" sensor, and the only available states are "Unavailable" and "Unknown".
> 
> That kinda seems like a bug? Is there an open issue in HA issues?

I do not believe there is an open issue at HA, would you recommend I create one?

Thank you for all your efforts @bdraco, I've followed the events here and appreciate the work you all do.

**Comment by bdraco** [comment]
> > > Currently HA users only have access to the "last_doorbell_ring" sensor, and the only available states are "Unavailable" and "Unknown".
> 
> > 
> 
> > That kinda seems like a bug? Is there an open issue in HA issues?
> 
> 
> 
> I do not believe there is an open issue at HA, would you recommend I create one?
> 
> 

Yes please. Hopefully someone knows what's wrong. 

I will have access to a UI doorbell next month when I'm visiting family so I can try to replicate it than if nobody has come up with a solution by than.


> 
> Thank you for all your efforts @bdraco, I've followed the events here and appreciate the work you all do.

**Comment by iCSpotRun** [comment]
@bdraco, I'm happy to send you G4 Doorbell to keep if you like. It's a non-Pro version but it should work. Let me know.

**Comment by darkrain-nl** [comment]
My doorbell works fine, it is in HA as `binary_sensor.g4_doorbell_pro_doorbell` and can be Occupied and Clear as states.

This still works after HA version 2024.6.2.

**Comment by iCSpotRun** [comment]
> My doorbell works fine, it is in HA as `binary_sensor.g4_doorbell_pro_doorbell` and can be Occupied and Clear as states.
> 
> This still works after HA version 2024.6.2.

That is an Occupancy sensor and not associated with the physical doorbell button on the device.

**Comment by bdraco** [comment]
> I'm happy to send you G4 Doorbell to keep if you like. It's a non-Pro version but it should work. Let me know.

I just logged into my family members's HA install and noticed that these sensors are disabled by default.  Once I enabled it, its showing the last ring as expected. I'm physically ~4000 miles away from where its located so I can't actually do any live testing though.

<img width="430" alt="Screenshot 2024-06-11 at 2 34 59 PM" src="https://github.com/uilibs/uiprotect/assets/663432/0d3ad036-4227-4e4d-9538-570d315a22cb">


Is there a chance the sensor is disabled in HA?

**Comment by darkrain-nl** [comment]
> > My doorbell works fine, it is in HA as `binary_sensor.g4_doorbell_pro_doorbell` and can be Occupied and Clear as states.
> > This still works after HA version 2024.6.2.
> 
> That is an Occupancy sensor and not associated with the physical doorbell button on the device.

I am sorry, then I have misunderstood the issue...

**Comment by iCSpotRun** [comment]
> > I'm happy to send you G4 Doorbell to keep if you like. It's a non-Pro version but it should work. Let me know.
> 
> I just logged into my family members's HA install and noticed that these sensors are disabled by default. Once I enabled it, its showing the last ring as expected.
> 
> <img alt="Screenshot 2024-06-11 at 2 34 59 PM" width="430" src="https://private-user-images.githubusercontent.com/663432/338722264-0d3ad036-4227-4e4d-9538-570d315a22cb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgxMzQ4MzQsIm5iZiI6MTcxODEzNDUzNCwicGF0aCI6Ii82NjM0MzIvMzM4NzIyMjY0LTBkM2FkMDM2LTQyMjctNGU0ZC05NTM4LTU3MGQzMTVhMjJjYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxMVQxOTM1MzRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NWM5YzFlYjc2MzlmZDExZGQwMDdiNzYwYzQ5MDk0MTAwZjM5YmFmYTc1YTkxNzgyNDM5ZWQ2OWRkZGM1NjA0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.01YI0Ydsd_sSDCtCfTJeWjvrGT6yoyRMYOnUgQYIp-c">
> Is there a chance the sensor is disabled in HA?

Correct, I did the same thing. The "Last Doorbell Ring" entity is disabled by default. It's the closest entity type I can find that meets my needs. But I'm afraid the only available states are Unknown and Unavailable; nothing that is associated with button presses on the device itself.

Is there something else in HA I have to enable?

**Comment by darkrain-nl** [comment]
On second thought, that sensor I mentioned is the one I use to detect the button being pressed... pretty sure that is what it is for... Is that not what you want?

**Comment by zhazell** [comment]
Yes, that binary sensor is what I also use for the doorbell press. However, it's just marked as an Occupancy sensor by default. 

![Screenshot 2024-06-11 at 1 43 51 PM](https://github.com/uilibs/uiprotect/assets/17417763/452bbca9-1b10-48b2-80be-3eafa1242c87)

**Comment by zhazell** [comment]
None of the other options really make sense. Seems HA should add a "button" type so it's either Pressed or Not-Pressed. 
![Screenshot 2024-06-11 at 1 45 47 PM](https://github.com/uilibs/uiprotect/assets/17417763/1040e847-4362-4bf7-bd1e-6a41b1dfba7d)

**Comment by darkrain-nl** [comment]
Is it not occupied as long as it's pressed? I cannot test that now as the family is already sleeping...

**Comment by iCSpotRun** [comment]
> On second thought, that sensor I mentioned is the one I use to detect the button being pressed... pretty sure that is what it is for... Is that not what you want?

It's exactly what I want, but that specific sensor doesn't function that way for me. It seems to flip between Cleared and Detected based on camera detection. This I've tested. Pressing the doorbell does not toggle this state.

I have the G4 Doorbell Pro PoE if this matters.

**Comment by iCSpotRun** [comment]
> Is it not occupied as long as it's pressed? I cannot test that now as the family is already sleeping...

I'm afraid not. Pressing the physical doorbell button does not toggle this state for me. Perhaps you can test when your family wakes? Haha

**Comment by darkrain-nl** [comment]
> > Is it not occupied as long as it's pressed? I cannot test that now as the family is already sleeping...
> 
> I'm afraid not. Pressing the physical doorbell button does not toggle this state for me. Perhaps you can test when your family wakes? Haha

Yes it does, I just tested it...
I made sure to not be in frame as well...

**Comment by iCSpotRun** [comment]
> > > Is it not occupied as long as it's pressed? I cannot test that now as the family is already sleeping...
> > 
> > 
> > I'm afraid not. Pressing the physical doorbell button does not toggle this state for me. Perhaps you can test when your family wakes? Haha
> 
> Yes it does, I just tested it... I made sure to not be in frame as well...


I do know that the specific sensor you are referring to changes states from Cleared to Occupied on camera movement in my case. So for my automation, the chime will be ringing through the house speakers when motion is detected. Very not wife approved.

I'll test again when I get home in case something has changed in recent updates.

**Reply by darkrain-nl** [reply]
Well if I walk up to the front door in frame of the camera the sensor does nothing in my case, when I press it it changes to Occupied for a few seconds and then back to Clear... I'm 100% sure it doesn't do anything with motion.

My device is a UVC G4 Doorbell Pro.

It would be really interesting to find out what is wrong here...

**Reply by iCSpotRun** [reply]
> Well if I walk up to the front door in frame of the camera the sensor does nothing in my case, when I press it it changes to Occupied for a few seconds and then back to Clear... I'm 100% sure it doesn't do anything with motion.
> 
> My device is a UVC G4 Doorbell Pro.
> 
> It would be really interesting to find out what is wrong here...

Ok, so I re-tested and good news. The doorbell binary sensor, while I believe incorrectly identified as an Occupancy sensor, is indeed mapped to the physical door bell button on the device. This was not how it behaved some time ago, but I'm glad that it's working properly now.

Thank you all for your feedback!

**Reply by bdraco** [reply]
I cleaned up the HA integration quite a bit this week so it should be much easier to add an event platform for the doorbell if anyone is interested in taking a stab at it

**Reply by iCSpotRun** [reply]
> I cleaned up the HA integration quite a bit this week so it should be much easier to add an event platform for the doorbell if anyone is interested in taking a stab at it 

You aren't kidding. That's a lot of changes you made haha.

I'll certainly test and get back to you after the weekend. I struggled listening to any protect events at the start of all this.

**Reply by bdraco** [reply]
https://github.com/home-assistant/core/pull/119746 has some more cleanups as well.  There is probably some more cleanups for has_entity_name that could be done to reduce the code a bit more

**Reply by bdraco** [reply]
I did some more cleanup over the weekend.  We still need to do entity translations but I think the entity base classes are lean enough that it won't cause any conflicts if someone does pickup adding an event entity

**Reply by darkrain-nl** [reply]
I lack the knowledge to do anything like that, but I do want to thank you for all this great work that you did.

**Comment by bdraco** [comment]
Since this is turning into a discussion, I'm going to go ahead and convert it
